File: backend/app/services/risk_engine.py
"""Transparent risk-prioritization engine.

Combines CVSS (severity) + EPSS (exploitation likelihood, with trend) + KEV (confirmed
exploitation, with recency) + exploit availability + temporal maturity to produce a
P1–P4 priority with human-readable reasons.

Each signal contributes independently to the score and produces a written justification.
The engine supports custom weight overrides so consumers can tune the prioritization.
"""

from datetime import datetime, timedelta, timezone

DEFAULT_WEIGHTS = {
    "cvss": 25,
    "epss": 20,
    "kev": 25,
    "exploit": 15,
    "temporal": 15,
}

THRESHOLDS = [
    (80, "Critical", "P1"),
    (60, "High", "P2"),
    (35, "Medium", "P3"),
]

RECOMMENDATIONS = {
    "P1": "Remediate immediately. Patch affected systems or apply mitigating controls without delay.",
    "P2": "Remediate urgently. Schedule patching within the current security cycle.",
    "P3": "Remediate in standard cycle. Include in next regular patch window.",
    "P4": "Monitor. Track for changes in exploitation status or trend shifts.",
}


def _config_weights(overrides: dict | None) -> dict:
    """Merge caller-provided weight overrides with defaults.

    Weights are normalized so the total == 100. Unknown keys are ignored.
    """
    weights = dict(DEFAULT_WEIGHTS)
    if overrides:
        for k in weights:
            if k in overrides and isinstance(overrides[k], (int, float)):
                weights[k] = max(0, float(overrides[k]))
    total = sum(weights.values())
    if total <= 0:
        return dict(DEFAULT_WEIGHTS)
    return {k: round(v / total * 100, 2) for k, v in weights.items()}


def _severity_from_cvss(score) -> str | None:
    if score is None:
        return None
    score = float(score)
    if score >= 9.0:
        return "Critical"
    if score >= 7.0:
        return "High"
    if score >= 4.0:
        return "Medium"
    return "Low"


def _parse_date(date_str: str | None):
    if not date_str:
        return None
    for fmt in ("%Y-%m-%d", "%Y-%m-%d %H:%M:%S", "%Y-%m-%dT%H:%M:%S"):
        try:
            return datetime.strptime(date_str[:19], fmt).replace(tzinfo=timezone.utc)
        except Exception:
            continue
    return None


def calculate_risk_assessment(
    cve_data: dict,
    affected_assets: list[dict] | None = None,
    weight_overrides: dict | None = None,
) -> dict:
    """Produce a full risk assessment from the assembled CVE data dict.

    The cve_data dict should contain:
        best_cvss: { score, severity, version }
        epss: { score, previous_score, percentile } | null
        kev: { vendor, product, date_added, due_date, ... } | null
        exploits: [{ url, name, stars }] | []
        published_date: str | null        (ISO date)
        modified_date: str | null

    Returns:
        {
            "score": 0-100,
            "level": "Critical" | "High" | "Medium" | "Low",
            "priority": "P1" | "P2" | "P3" | "P4",
            "signals": [ { name, label, value, max, weight, reason? }, ... ],
            "reasons": [ str, ... ],
            "recommendation": str,
            "trends": { "epss_delta", "kev_recent" },
        }
    """
    weights = _config_weights(weight_overrides)
    signals = []
    reasons = []
    total = 0.0

    cvss = cve_data.get("best_cvss") or {}
    cvss_score = cvss.get("score") or 0
    severity = cvss.get("severity") or "Unknown"

    epss = cve_data.get("epss") or {}
    epss_score = epss.get("score") or 0
    prev_epss = epss.get("previous_score")
    epss_delta = (epss_score - prev_epss) if prev_epss is not None else None

    kev = cve_data.get("kev")
    has_kev = kev is not None

    exploits = cve_data.get("exploits") or []
    has_exploit = len(exploits) > 0
    top_stars = exploits[0].get("stars", 0) if exploits else 0

    affects = []
    if affected_assets:
        affects = affected_assets

    # --- 1. CVSS severity (weighted) ---
    cvss_max = weights["cvss"]
    if cvss_score:
        cvss_points = round((float(cvss_score) / 10) * cvss_max)
    else:
        cvss_points = 0
    total += cvss_points
    signals.append({
        "name": "cvss",
        "label": f"CVSS {severity} ({cvss_score}/10)",
        "value": cvss_points,
        "max": cvss_max,
        "weight": f"{weights['cvss']:.0f}%",
    })
    if cvss_score >= 9.0:
        reasons.append(f"Critical technical severity (CVSS {cvss_score:.1f})")
    elif cvss_score >= 7.0:
        reasons.append(f"High technical severity (CVSS {cvss_score:.1f})")
    elif cvss_score >= 4.0:
        reasons.append(f"Moderate technical severity (CVSS {cvss_score:.1f})")
    else:
        reasons.append("Low technical severity")

    # --- 2. EPSS exploitation probability + trend (weighted) ---
    epss_max = weights["epss"]
    epss_points = round(float(epss_score) * epss_max)
    total += epss_points
    pct_label = f"{float(epss_score) * 100:.1f}%"
    trend_arrow = ""
    if epss_delta is not None:
        trend_arrow = " ↑" if epss_delta > 0.01 else (" ↓" if epss_delta < -0.01 else " →")
    if epss_score >= 0.7:
        label_rank = "Very High"
        reasons.append(f"Very high exploitation probability (EPSS {pct_label})")
    elif epss_score >= 0.4:
        label_rank = "High"
        reasons.append(f"High exploitation probability (EPSS {pct_label})")
    elif epss_score >= 0.1:
        label_rank = "Moderate"
        reasons.append(f"Moderate exploitation probability (EPSS {pct_label})")
    else:
        label_rank = "Low"
        reasons.append(f"Low exploitation probability (EPSS {pct_label})")
    signals.append({
        "name": "epss",
        "label": f"EPSS {pct_label} — {label_rank}{trend_arrow}",
        "value": epss_points,
        "max": epss_max,
        "weight": f"{weights['epss']:.0f}%",
    })
    if epss_delta is not None:
        delta_pct = abs(epss_delta) * 100
        if epss_delta > 0.01:
            signals[-1]["reason"] = f"EPSS rose {delta_pct:.1f} pts vs previous value"
            reasons.append(f"EPSS trending upward (+{delta_pct:.1f}%) — exploitation likelihood increasing")
        elif epss_delta < -0.01:
            reasons.append(f"EPSS trending downward ({delta_pct:.1f}%) — likelihood decreasing")

    # --- 3. KEV confirmed exploitation (weighted) + recency maturity ---
    kev_max = weights["kev"]
    kev_points = kev_max if has_kev else 0
    kev_recent = False
    if has_kev:
        total += kev_points
        kev_reason = "Confirmed exploitation in the wild (CISA KEV catalog)"
        date_added = kev.get("date_added")
        added_dt = _parse_date(str(date_added)) if date_added else None
        if added_dt and datetime.now(timezone.utc) - added_dt <= timedelta(days=90):
            kev_recent = True
            kev_reason += " — added recently (within 90 days)"
            reasons.append(kev_reason)
        else:
            reasons.append(kev_reason)
        if kev.get("known_ransomware_campaign_use"):
            reasons.append("Exploited in known ransomware campaigns")
    signals.append({
        "name": "kev",
        "label": ("CISA KEV — Confirmed exploited (recent)" if kev_recent and has_kev
                  else "CISA KEV — Confirmed exploited" if has_kev else "Not in CISA KEV"),
        "value": kev_points,
        "max": kev_max,
        "weight": f"{weights['kev']:.0f}%",
    })

    # --- 4. Public exploit availability (weighted) ---
    exploit_max = weights["exploit"]
    if has_exploit and top_stars > 100:
        exploit_points = round(exploit_max)
        reasons.append(f"Widely available exploit code ({top_stars} stars)")
    elif has_exploit:
        exploit_points = round(exploit_max * 0.66)
        reasons.append("Public exploit code available")
    else:
        exploit_points = 0
    total += exploit_points
    signals.append({
        "name": "exploit",
        "label": f"{len(exploits)} exploit source{'s' if len(exploits) != 1 else ''}" if has_exploit else "No public exploits",
        "value": exploit_points,
        "max": exploit_max,
        "weight": f"{weights['exploit']:.0f}%",
    })

    # --- 5. Temporal maturity (weighted) — fresh disclosure in active attack window ---
    temporal_max = weights["temporal"]
    temporal_points = 0
    age_days = None
    published = _parse_date(str(cve_data.get("published_date"))) if cve_data.get("published_date") else None
    now = datetime.now(timezone.utc)
    if published:
        age_days = (now - published).days
        if age_days <= 30:
            temporal_points = round(temporal_max)
            reasons.append(f"Recently disclosed ({age_days}d old) — active exploitation window")
        elif age_days <= 120:
            temporal_points = round(temporal_max * 0.66)
            reasons.append(f"Moderately recent disclosure ({age_days}d old)")
        elif age_days <= 365 and (has_kev or has_exploit):
            temporal_points = round(temporal_max * 0.33)
            reasons.append(f"Disclosed within past year and actively targeted")
    total += temporal_points
    signals.append({
        "name": "temporal",
        "label": f"{age_days}d old" if published else "Disclosure age unknown",
        "value": temporal_points,
        "max": temporal_max,
        "weight": f"{weights['temporal']:.0f}%",
    })

    # --- Final score ---
    score = min(100, round(total))

    level, priority = "Low", "P4"
    for threshold, lvl, pri in THRESHOLDS:
        if score >= threshold:
            level, priority = lvl, pri
            break

    return {
        "score": score,
        "level": level,
        "priority": priority,
        "signals": signals,
        "reasons": reasons,
        "recommendation": RECOMMENDATIONS[priority],
        "trends": {
            "epss_delta": epss_delta,
            "kev_recent": kev_recent,
        },
    }

File: src/utils/riskEngine.js
/**
 * Transparent client-side risk-prioritization engine.
 *
 * Combines CVSS (severity) + EPSS (exploitation likelihood, with trend) +
 * KEV (confirmed exploitation, with recency) + exploit availability +
 * temporal maturity to produce a P1–P4 priority with human-readable reasons.
 *
 * 100% browser-executable — no backend or database required.
 */

const DEFAULT_WEIGHTS = {
  cvss: 25,
  epss: 20,
  kev: 25,
  exploit: 15,
  temporal: 15,
}

const THRESHOLDS = [
  [80, 'Critical', 'P1'],
  [60, 'High', 'P2'],
  [35, 'Medium', 'P3'],
]

const RECOMMENDATIONS = {
  P1: 'Remediate immediately. Patch affected systems or apply mitigating controls without delay.',
  P2: 'Remediate urgently. Schedule patching within the current security cycle.',
  P3: 'Remediate in standard cycle. Include in next regular patch window.',
  P4: 'Monitor. Track for changes in exploitation status or trend shifts.',
}

function normalizeWeights(overrides) {
  const weights = { ...DEFAULT_WEIGHTS }
  if (overrides && typeof overrides === 'object') {
    for (const k of Object.keys(weights)) {
      if (typeof overrides[k] === 'number') {
        weights[k] = Math.max(0, overrides[k])
      }
    }
  }
  const total = Object.values(weights).reduce((a, b) => a + b, 0)
  if (total <= 0) return { ...DEFAULT_WEIGHTS }
  const out = {}
  for (const [k, v] of Object.entries(weights)) {
    out[k] = (v / total) * 100
  }
  return out
}

function parseDate(dateStr) {
  if (!dateStr) return null
  const d = new Date(dateStr)
  return isNaN(d.getTime()) ? null : d
}

/**
 * Produce a full risk assessment from assembled CVE data dict.
 */
export function calculateRiskAssessment(cveData, weightOverrides = null) {
  const weights = normalizeWeights(weightOverrides)
  const signals = []
  const reasons = []
  let total = 0

  const cvss = cveData.best_cvss || {}
  const cvssScore = typeof cvss.score === 'number' ? cvss.score : parseFloat(cvss.score) || 0
  const severity = cvss.severity || 'Unknown'

  const epss = cveData.epss || {}
  const epssScore = typeof epss.score === 'number' ? epss.score : parseFloat(epss.score) || 0
  const prevEpss = typeof epss.previous_score === 'number' ? epss.previous_score : null
  const epssDelta = prevEpss !== null ? epssScore - prevEpss : null

  const kev = cveData.kev
  const hasKev = Boolean(kev)

  const exploits = Array.isArray(cveData.exploits) ? cveData.exploits : []
  const hasExploit = exploits.length > 0
  const topStars = hasExploit ? (exploits[0].stars || 0) : 0

  // 1. CVSS severity
  const cvssMax = weights.cvss
  const cvssPoints = cvssScore ? Math.round((cvssScore / 10) * cvssMax) : 0
  total += cvssPoints
  signals.push({
    name: 'cvss',
    label: `CVSS ${severity} (${cvssScore.toFixed(1)}/10)`,
    value: cvssPoints,
    max: cvssMax,
    weight: `${Math.round(weights.cvss)}%`,
  })
  if (cvssScore >= 9.0) {
    reasons.push(`Critical technical severity (CVSS ${cvssScore.toFixed(1)})`)
  } else if (cvssScore >= 7.0) {
    reasons.push(`High technical severity (CVSS ${cvssScore.toFixed(1)})`)
  } else if (cvssScore >= 4.0) {
    reasons.push(`Moderate technical severity (CVSS ${cvssScore.toFixed(1)})`)
  } else {
    reasons.push('Low technical severity')
  }

  // 2. EPSS exploitation probability
  const epssMax = weights.epss
  const epssPoints = Math.round(epssScore * epssMax)
  total += epssPoints
  const pctLabel = `${(epssScore * 100).toFixed(1)}%`
  let trendArrow = ''
  if (epssDelta !== null) {
    trendArrow = epssDelta > 0.01 ? ' ↑' : (epssDelta < -0.01 ? ' ↓' : ' →')
  }
  let labelRank = 'Low'
  if (epssScore >= 0.7) {
    labelRank = 'Very High'
    reasons.push(`Very high exploitation probability (EPSS ${pctLabel})`)
  } else if (epssScore >= 0.4) {
    labelRank = 'High'
    reasons.push(`High exploitation probability (EPSS ${pctLabel})`)
  } else if (epssScore >= 0.1) {
    labelRank = 'Moderate'
    reasons.push(`Moderate exploitation probability (EPSS ${pctLabel})`)
  } else {
    reasons.push(`Low exploitation probability (EPSS ${pctLabel})`)
  }

  const epssSignal = {
    name: 'epss',
    label: `EPSS ${pctLabel} — ${labelRank}${trendArrow}`,
    value: epssPoints,
    max: epssMax,
    weight: `${Math.round(weights.epss)}%`,
  }
  if (epssDelta !== null) {
    const deltaPct = (Math.abs(epssDelta) * 100).toFixed(1)
    if (epssDelta > 0.01) {
      epssSignal.reason = `EPSS rose ${deltaPct} pts vs previous value`
      reasons.push(`EPSS trending upward (+${deltaPct}%) — exploitation likelihood increasing`)
    } else if (epssDelta < -0.01) {
      reasons.push(`EPSS trending downward (-${deltaPct}%) — likelihood decreasing`)
    }
  }
  signals.push(epssSignal)

  // 3. KEV confirmed exploitation
  const kevMax = weights.kev
  const kevPoints = hasKev ? kevMax : 0
  let kevRecent = false
  if (hasKev) {
    total += kevPoints
    let kevReason = 'Confirmed exploitation in the wild (CISA KEV catalog)'
    const dateAdded = kev.date_added || kev.dateAdded
    const addedDt = parseDate(dateAdded)
    if (addedDt && (Date.now() - addedDt.getTime()) <= 90 * 24 * 60 * 60 * 1000) {
      kevRecent = true
      kevReason += ' — added recently (within 90 days)'
    }
    reasons.push(kevReason)
    if (kev.known_ransomware_campaign_use || kev.knownRansomwareCampaignUse === 'Known') {
      reasons.push('Exploited in known ransomware campaigns')
    }
  }
  signals.push({
    name: 'kev',
    label: kevRecent && hasKev
      ? 'CISA KEV — Confirmed exploited (recent)'
      : (hasKev ? 'CISA KEV — Confirmed exploited' : 'Not in CISA KEV'),
    value: kevPoints,
    max: kevMax,
    weight: `${Math.round(weights.kev)}%`,
  })

  // 4. Public exploit availability
  const exploitMax = weights.exploit
  let exploitPoints = 0
  if (hasExploit && topStars > 100) {
    exploitPoints = Math.round(exploitMax)
    reasons.push(`Widely available exploit code (${topStars} stars)`)
  } else if (hasExploit) {
    exploitPoints = Math.round(exploitMax * 0.66)
    reasons.push('Public exploit code available')
  }
  total += exploitPoints
  signals.push({
    name: 'exploit',
    label: hasExploit ? `${exploits.length} exploit source${exploits.length !== 1 ? 's' : ''}` : 'No public exploits',
    value: exploitPoints,
    max: exploitMax,
    weight: `${Math.round(weights.exploit)}%`,
  })

  // 5. Temporal maturity
  const temporalMax = weights.temporal
  let temporalPoints = 0
  let ageDays = null
  const publishedDt = parseDate(cveData.published_date)
  if (publishedDt) {
    ageDays = Math.floor((Date.now() - publishedDt.getTime()) / (24 * 60 * 60 * 1000))
    if (ageDays <= 30) {
      temporalPoints = Math.round(temporalMax)
      reasons.push(`Recently disclosed (${ageDays}d old) — active exploitation window`)
    } else if (ageDays <= 120) {
      temporalPoints = Math.round(temporalMax * 0.66)
      reasons.push(`Moderately recent disclosure (${ageDays}d old)`)
    } else if (ageDays <= 365 && (hasKev || hasExploit)) {
      temporalPoints = Math.round(temporalMax * 0.33)
      reasons.push('Disclosed within past year and actively targeted')
    }
  }
  total += temporalPoints
  signals.push({
    name: 'temporal',
    label: ageDays !== null ? `${ageDays}d old` : 'Disclosure age unknown',
    value: temporalPoints,
    max: temporalMax,
    weight: `${Math.round(weights.temporal)}%`,
  })

  const score = Math.min(100, Math.max(0, Math.round(total)))
  let level = 'Low'
  let priority = 'P4'
  for (const [threshold, lvl, pri] of THRESHOLDS) {
    if (score >= threshold) {
      level = lvl
      priority = pri
      break
    }
  }

  return {
    score,
    level,
    priority,
    signals,
    reasons,
    recommendation: RECOMMENDATIONS[priority],
    trends: {
      epss_delta: epssDelta,
      kev_recent: kevRecent,
    },
  }
}
